tests/test_love_producer_base.py

```diff
@@ -0,0 +1,376 @@
+# This file is part of ts_salobj.
+#
+# Developed for the Rubin Observatory Telescope and Site System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import json
+import asyncio
+import unittest
+import logging
+
+import numpy as np
+
+from astropy.time import Time
+
+from love.producer import LoveProducerBase
+from love.producer.test_utils import cancel_task
+
+
+class TestLoveProducerBase(unittest.IsolatedAsyncioTestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.log = logging.getLogger(__name__)
+        cls.standard_timeout = 5
+        cls.time_pooling = 0.1
+
+    async def asyncSetUp(self):
+        self.producer = LoveProducerBase()
+        self.messages_received = []
+        self._run_data_generator_loop = False
+        self.data_generator_task = asyncio.Future()
+        self.data_generator_task.set_result(True)
```

Comment:
How does this work?

Reply:
The `data_generator_task` attribute is used to store a task later that will be generating data (created in `setup_data_generation_loop`). Here it is just creating a future and setting the result of the future. 

The reason for this is that not all unit tests will call `setup_data_generation_loop`. When cleaning up the unit tests I need to make sure I interrupt any pending task. So, I check if this task is done and deal with the condition when it is not. For those tests that don't use the data generation loop, they will simply ignore this cleanup step, because the task is already completed.

Reply:
I understand now! thanks